AtelierTomato.Markov/AtelierTomato.Markov.Renderer/SentenceRenderer.cs

```diff
@@ -0,0 +1,44 @@
+﻿
+using System.Text.RegularExpressions;
+
+namespace AtelierTomato.Markov.Renderer
+{
+	public class SentenceRenderer
+	{
+		protected readonly Regex renderEmojiRegex = new Regex(@"e:((?:(?!:).)+):", RegexOptions.Compiled);
+		private readonly Regex attachToPreviousWord = new Regex(@"(?: |^)([.\}\)\];:]|[?!,]+)(?: |$)", RegexOptions.Compiled);
+		private readonly Regex attachToNextWord = new Regex(@"(?: |^)([#\[\{\(])(?: |$)", RegexOptions.Compiled);
+		private readonly Regex attachQuotes = new Regex(@"(?: |^)("")(?: |$)", RegexOptions.Compiled);
+		private readonly Regex attachContractions = new Regex(@"(?: )('\w+)", RegexOptions.Compiled);
+		private readonly Regex attachPluralContractions = new Regex(@"(?: )(')(?: |$)", RegexOptions.Compiled);
+		private readonly Regex attachDashes = new Regex(@"(\w+-)(?: )(\w+)", RegexOptions.Compiled);
+		private readonly Regex attachQuoteArrow = new Regex(@"(?:^)(>)(?: )", RegexOptions.Compiled);
+
+		public virtual string Render(string text)
+		{
+			text = RenderDetachCharacters(text);
+			text = RenderEmojis(text);
+			return text;
+		}
+
+		private string RenderDetachCharacters(string text)
+		{
+			text = attachToPreviousWord.Replace(text, m => m.Groups[1] + " ");

```

Comment:
i think this adds unnecessary spaces at the end of the whole text, maybe we can do m.Groups[2] (similar thing in next line)

Reply:
oh nvm we use ?: groups for some reason, i guess this is ok then